contracts/rsv/Reserve.sol

```diff
@@ -0,0 +1,367 @@
+pragma solidity ^0.5.8;
+
+import "../zeppelin/token/ERC20/IERC20.sol";
+import "../zeppelin/math/SafeMath.sol";
+import "../ownership/Ownable.sol";
+import "./ReserveEternalStorage.sol";
+
+/**
+ * @title An interface representing a contract that calculates transaction fees
+ */
+ interface ITXFee {
+    function calculateFee(address, uint256) external returns (uint256);
+ }
+
+/**
+ * @title The Reserve Token
+ * @dev An ERC-20 token with minting, burning, pausing, and user freezing.
+ * Based on OpenZeppelin's [implementation](https://github.com/OpenZeppelin/openzeppelin-solidity/blob/41aa39afbc13f0585634061701c883fe512a5469/contracts/token/ERC20/ERC20.sol).
+ *
+ * Non-constant-sized data is held in ReserveEternalStorage,
+ * to facilitate potential future upgrades.
+ */
+contract Reserve is IERC20, Ownable {
+    using SafeMath for uint256;
+
+
+    // Non-constant-sized data
+    ReserveEternalStorage internal data;
+
+    // TX Fee helper contract
+    ITXFee public txFee;
+
+    // Basic token data
+    string public name = "Reserve";
+    string public symbol = "RSV";
+    uint8 public constant decimals = 18;
+    uint256 public totalSupply;
+    uint256 public maxSupply;
+
+    // Paused data
+    bool public paused;
+
+    // Auth roles
+    address public minter;
+    address public pauser;
+    address public freezer;
+    address public feeRecipient;
+
+    // Auth role change events
+    event MinterChanged(address indexed newMinter);
+    event PauserChanged(address indexed newPauser);
+    event FreezerChanged(address indexed newFreezer);
+    event FeeRecipientChanged(address indexed newFeeRecipient);
+    event MaxSupplyChanged(uint256 indexed newMaxSupply);
+
+    // Pause events
+    event Paused(address indexed account);
+    event Unpaused(address indexed account);
+
+    // Name change event
+    event NameChanged(string newName, string newSymbol);
+
+    // Law enforcement events
+    event Frozen(address indexed freezer, address indexed account);
+    event Unfrozen(address indexed freezer, address indexed account);
+    event Wiped(address indexed freezer, address indexed wiped);
+
+
+    /// Initialize critical fields.
+    constructor() public {
+        data = new ReserveEternalStorage(msg.sender);
+        txFee = ITXFee(address(0)); // I'm not sure if this will fail here or later. If it fails here, then we'll need a different design
```

Comment:
I would like to determine whether this will fail here or later before we approve of this code. :)

Reply:
It doesn't fail there